README.md
# MiniUrll

## Overview

MiniUrll is a Web Application for shortening URLs.

-   _Single Page Application_ with most of the functionalies that a URL shortener.
-   _FetchAPI_ for requesting the server.
-   _Passwordless_ login system.
-   _JSON_ for sending requests and responses to-and-fro server.

### Built with

-   HTML
-   CSS
-   JavaScript
-   Flask
-   PostgreSQL

### Screenshots

![Desktop Preview](./screenshots/desktop_preview.png)

### Live Demo

[https://miniurll.herokuapp.com](https://miniurll.herokuapp.com)

### Useful Resources

[MDN Docs](https://developer.mozilla.org/en-US/) : Best documentation for web technologies including HTML, CSS, JavaScript and many more.

[Flask Documentation](https://flask.palletsprojects.com/en/2.1.x/) : Official Flask documentation.

[W3SCHOOLS](https://www.w3schools.com/) : Best resource to get information on everything about web.

## Author

-   Website - [Jatin Nagar](https://itsjatinnagar.blogspot.com)
-   Twitter - [@itsjatinnagar](https://www.twitter.com/itsjatinnagar)


CHANGELOG.md
# CHANGE LOGS

## VERSION 2.4 (2022-05-31)

### Fixed

-   Issue [#6](https://github.com/itsjatinnagar/url-shortener/issues/6)

## VERSION 2.3 (2022-05-30)

### Fixed

-   Issue [#5](https://github.com/itsjatinnagar/url-shortener/issues/5)

## VERSION 2.2 (2022-05-17)

### Fixed

-   Unequal heights of the Feature Card

## VERSION 2.1 (2022-05-14)

### Fixed

-   Issue [#4](https://github.com/itsjatinnagar/url-shortener/issues/4)

## VERSION 2.0 (2022-05-12)

### Added

-   Passwordless Login functionality
-   Showing form based upon user authentication.

### Changed

-   Database Schema
-   Storing user information on server.
-   Accessing shortened urls using user information.

### Removed

-   Using cookies for storing short_codes.

## VERSION 1.2 (2022-05-04)

### Fixed

-   Issue [#2](https://github.com/itsjatinnagar/url-shortener/issues/2)

## VERSION 1.1 (2022-05-01)

### Fixed

-   Issue [#1](https://github.com/itsjatinnagar/url-shortener/issues/1)

## VERSION 1.0 (2022-04-30)

### Added

-   Final fixes
-   Copy To Clipboard and Button Disabling
-   Handling Server Response
-   Popup Toggle
-   Get and Set Cookies
-   JINJA Templating
-   Created: Redirect route
-   Fixes: server error handling
-   Designed: Popup modal
-   Check: Unique short-codes
-   Defined: Table INSERT
-   Database setup
-   Defined: ShortCode generating function
-   AJAX/Fetch: post request
-   Created: Url shortener route
-   Defined: Validator function
-   fixed filepaths
-   Setup: flask app
-   Defined: JS function
-   Fixes: Template and Design
-   Designed: Links Section
-   Designed: Footer
-   Designed: Features Section
-   Designed: Shorten Form
-   Designed: Landing Section
-   Designed: Navigation bar
-   initial commit


app.py
from datetime import datetime
import os
from flask import Flask, jsonify, redirect, render_template, request, session, url_for

from model.helpers import email_auth_code, generate_auth_code, generate_short_code
from model.sql import insertURL, insertUser, read_all, read_long_url, read_user_id
from model.validators import validateAuthCode, validateEmail, validateURL

app = Flask(__name__, static_folder='assets')
app.secret_key = os.environ['SESSION_KEY']


@app.route('/')
def index():
    try:
        identifiers = session['id']
    except:
        identifiers = None

    if identifiers is None:
        return render_template('index.html', isAuth=False)
    else:
        records = read_all(identifiers)
        return render_template('index.html', isAuth=True, url_list=records, host=request.host_url)


@app.route('/mail-code', methods=['POST'])
def mail_auth_code():
    email = request.json['email'].strip()
    validate = validateEmail(email)
    if validate is None:
        session['email'] = email
        session['auth_code'] = generate_auth_code()
        isSuccess = email_auth_code(session['email'], session['auth_code'])
        if isSuccess:
            return jsonify(status=200, message='Verification Code Sent')
        else:
            return jsonify(status=500, message='Something Went Wrong')
    else:
        return jsonify(status=400, message=validate)


@app.route('/login', methods=['POST'])
def login():
    user_code = request.json['code'].strip()
    validate = validateAuthCode(user_code, session['auth_code'])
    if validate is None:
        user_id = read_user_id(session['email'])
        if user_id is False:
            isSuccess = False
        elif user_id is None:
            isSuccess = insertUser(session['email'])
        else:
            session.pop('auth_code')
            session['id'] = user_id[0]
            return jsonify(status=200, message='Logged In Successfully')

        if isSuccess is False:
            return jsonify(status=500, message='Something Went Wrong')
        else:
            session.pop('auth_code')
            session['id'] = read_user_id(session['email'])[0]
            return jsonify(status=200, message='Logged In Successfully')
    else:
        return jsonify(status=400, message=validate)


@app.route('/logout')
def logout():
    session.clear()
    return redirect(url_for('index'))


@app.route('/shorten', methods=['POST'])
def shorten():
    long_url = request.json['long_url'].strip()
    validate = validateURL(long_url)
    if validate is None:
        short_code = generate_short_code()
        if short_code is None:
            isSuccess = False
        else:
            date = datetime.now().strftime('%d-%m-%Y %H:%M:%S')
            # URL Insert
            isSuccess = insertURL(session['id'], short_code, long_url, date)
        if isSuccess is False:
            return jsonify(status=500, message='Something Went Wrong')
        else:
            return jsonify(status=200, message='URL Shortened Successfully', long_url=long_url, short_url=short_code)
    else:
        return jsonify(status=400, message=validate)


@app.route('/<short_code>')
def redirection(short_code):
    long_url = read_long_url(short_code)
    if long_url is None:
        return redirect(url_for('index'))

    return redirect(long_url[0])


model/sql.py
import os
import psycopg2


def connect():
    try:
        connection = psycopg2.connect(
            host=os.environ['DB_HOST'],
            database=os.environ['DB_NAME'],
            user=os.environ['DB_USER'],
            password=os.environ['DB_PASS'])
        return connection
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')


def insertUser(email):
    conn = connect()
    if conn is None:
        return False

    cursor = conn.cursor()
    query = 'INSERT INTO users (email) VALUES (%s)'
    values = (email,)
    try:
        cursor.execute(query, values)
        conn.commit()
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    cursor.close()
    conn.close()


def insertURL(uid, short_code, long_url, datetime):
    conn = connect()
    if conn is None:
        return False

    cursor = conn.cursor()
    query = 'INSERT INTO url VALUES (%s, %s, %s, %s)'
    values = (uid, short_code, long_url, datetime,)
    try:
        cursor.execute(query, values)
        conn.commit()
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    cursor.close()
    conn.close()


def check_code(short_code):
    conn = connect()
    if conn is None:
        return None

    cursor = conn.cursor()
    query = 'SELECT * FROM url where short_code = %s'
    values = (short_code,)
    try:
        cursor.execute(query, values)
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    queryResult = cursor.fetchall()

    cursor.close()
    conn.close()
    return True if len(queryResult) == 0 else False


def read_user_id(email):
    conn = connect()
    if conn is None:
        return False

    cursor = conn.cursor()
    query = 'SELECT id FROM users WHERE email = %s'
    values = (email,)
    try:
        cursor.execute(query, values)
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    queryResult = cursor.fetchone()

    cursor.close()
    conn.close()
    return queryResult


def read_long_url(short_code):
    conn = connect()
    if conn is None:
        return False

    cursor = conn.cursor()
    query = 'SELECT long_url FROM url where short_code = %s'
    values = (short_code,)
    try:
        cursor.execute(query, values)
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    queryResult = cursor.fetchone()

    cursor.close()
    conn.close()

    return queryResult


def read_all(uid):
    conn = connect()
    if conn is None:
        return False

    cursor = conn.cursor()
    query = 'SELECT long_url,short_code FROM url where uid = %s ORDER BY expiration DESC'
    values = (uid,)
    try:
        cursor.execute(query, values)
    except (Exception, psycopg2.Error) as error:
        print(f'Error: {error}')
        conn.close()
        return False

    queryResult = cursor.fetchall()

    cursor.close()
    conn.close()

    return queryResult


model/initDB.py
from sql import connect


conn = connect()
cursor = conn.cursor()

# Drop (IF)Existing Tables
cursor.execute('DROP TABLE IF EXISTS users, url, expired')

# Create Tables
cursor.execute('CREATE TABLE users('
               'id SERIAL,'
               'email VARCHAR(320) UNIQUE NOT NULL,'
               'PRIMARY KEY (id))'
               )

cursor.execute('CREATE TABLE url('
               'uid INTEGER NOT NULL,'
               'short_code CHAR(6) PRIMARY KEY,'
               'long_url VARCHAR(2000) NOT NULL,'
               'expiration CHAR(19) NOT NULL,'
               'FOREIGN KEY (uid) REFERENCES users(id))'
               )

cursor.execute('CREATE TABLE expired('
               'short_code CHAR(6) PRIMARY KEY,'
               'long_url VARCHAR(2000) NOT NULL)'
               )

conn.commit()

cursor.close()
conn.close()


model/helpers.py
from email.message import EmailMessage
import os
import smtplib
import string
import random

from flask import session

from model.sql import check_code


def generate_short_code():
    obsolete = 'lIO0'
    allowed = ''.join([('' if c in obsolete else c)
                      for c in string.ascii_letters + string.digits])
    short_code = ''
    while True:
        short_code = ''.join(random.choices(allowed, k=6))
        checkResult = check_code(short_code)
        if checkResult is True:
            break
        elif checkResult is None:
            return None
    return short_code


def generate_auth_code():
    return ''.join(random.choices(string.digits, k=6))


def email_auth_code(email: str, code):
    user = os.environ['EMAIL_ADDR']
    password = os.environ['EMAIL_PASS']
    message = EmailMessage()
    message['From'] = user
    message['To'] = email
    message['Subject'] = 'Your miniurl.cf sign in code'
    # Fallback Content
    message.set_content(
        f"Hey {email.split('@')[0]}!\n\nYour sign in code is: {code}\n\nThanks,\nThe MiniUrl Team")
    # HTML Content
    message.add_alternative(f"""\
    <!DOCTYPE html>
    <html lang="en">
    <body>
        <p style="font-size:18px;font-weight:bold;">Hey {email.split('@')[0]}!</p>
        <p style="font-size:20px;text-align:center;">Your sign in code is: <strong style="color:#2acf5f;">{code}</strong></h2>
        <p>Thanks,</p>
        <p>The MiniUrl Team</p>
    </body>
    </html>
    """, subtype='html')
    with smtplib.SMTP_SSL('smtp.gmail.com', 465) as smtp:
        try:
            smtp.login(user, password)
            smtp.send_message(message)
        except smtplib.SMTPException as error:
            print(f'Auth Error: {error}')
            return False
        return True


model/validators.py
import re


def validateURL(long_url):
    if len(long_url) <= 0:
        return 'Please, provide a valid URL'
    if len(long_url) > 2000:
        return 'URL too looonng'
    expression = '[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)'
    if re.search(expression, long_url) is None:
        return 'It is not a valid URL'

    return None


def validateEmail(email):
    if len(email) <= 0:
        return 'Please, provide your Email'
    expression = '^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$'
    if re.search(expression, email) is None:
        return 'It is not a valid Email'

    return None


def validateAuthCode(user_input, auth_code):
    if len(user_input) <= 0:
        return 'Please, enter the code sent to you'
    elif user_input != auth_code:
        return 'Invalid sign in code'

    return None


assets/js/script.js
const linksWrapper = document.getElementById("links-wrapper"),
    linkWrappers = document.getElementsByClassName("link-wrapper"),
    popup = document.getElementById("popup"),
    popupIcon = document.querySelector("#popup #icon i"),
    popupMessage = document.querySelector("#popup #message"),
    urlFieldWrapper = document.getElementById("field-url"),
    emailFieldWrapper = document.getElementById("field-email"),
    codeFieldWrapper = document.getElementById("field-code"),
    submitBtn = document.getElementById("submit-btn"),
    urlInput = document.getElementById("url"),
    emailInput = document.getElementById("email"),
    codeInput = document.getElementById("code"),
    form = document.getElementById("form");

const fetchRequest = (route, data) =>
    fetch(route, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
    })
        .then((response) => response.json())
        .then(data)
        .catch((error) => console.log("Error: " + error));

const checkResponse = (response, fieldWrapper) => {
    if (response.status === 200) {
        popupToggle("success", "fa-circle-check", response.message);
        return true;
    } else if (response.status === 400) {
        fieldWrapper.classList.add("error");
        fieldWrapper.children[1].textContent = response.message;
        setTimeout(() => {
            fieldWrapper.classList.remove("error");
            fieldWrapper.children[1].textContent = "";
        }, 4000);
        return false;
    } else {
        popupToggle("error", "fa-circle-exclamation", response.message);
        return false;
    }
};

const popupToggle = (className, iconClass, message) => {
    popup.classList.add(className);
    popupIcon.classList.add(iconClass);
    popupMessage.textContent = message;

    setTimeout(() => {
        popup.classList.remove(className);
        popup.addEventListener("transitionend", () => {
            popupIcon.classList.remove(iconClass);
            popupMessage.textContent = "";
        });
    }, 4000);
};

const createCardElement = (long_url, short_url) => {
    const linkWrapperDiv = document.createElement("div");
    linkWrapperDiv.classList.add("link-wrapper");
    const paraLongUrl = document.createElement("p");
    paraLongUrl.classList.add("long-link");
    paraLongUrl.innerText = long_url;
    const paraShortUrl = document.createElement("p");
    paraShortUrl.classList.add("short-link");
    paraShortUrl.innerText = window.location.origin + "/" + short_url;
    const btn = document.createElement("button");
    btn.classList.add(...["btn", "copy-btn"]);
    btn.innerText = "Copy";

    linkWrapperDiv.append(...[paraLongUrl, paraShortUrl, btn]);

    linksWrapper.insertBefore(linkWrapperDiv, linksWrapper.children[0]);

    copyToClipboard();
};

const copyToClipboard = () => {
    for (const wrapper of linkWrappers) {
        // Copy: Short url to Clipboard
        const copyBtn = wrapper.children[2];
        copyBtn.addEventListener("click", () => {
            navigator.clipboard.writeText(wrapper.children[1].innerText);

            copyBtn.innerText = "Copied!";
            copyBtn.classList.add("active");

            setTimeout(() => {
                copyBtn.innerText = "Copy";
                copyBtn.classList.remove("active");
            }, 3000);
        });
    }
};

let isCodeSent = false,
    isAuthenticated = false,
    isShortened = false,
    urlObject;
form.addEventListener("submit", async (e) => {
    e.preventDefault();
    submitBtn.setAttribute("disabled", "true");

    const submitBtnClasses = submitBtn.classList;
    if (submitBtnClasses.contains("send-code-btn")) {
        let response = await fetchRequest("/mail-code", {
            email: emailInput.value,
        });
        isCodeSent = checkResponse(response, emailFieldWrapper);
    } else if (submitBtnClasses.contains("sign-in-btn") && isCodeSent) {
        let response = await fetchRequest("/login", {
            code: codeInput.value,
        });
        isAuthenticated = checkResponse(response, codeFieldWrapper);
    } else if (submitBtnClasses.contains("shorten-btn")) {
        let response = await fetchRequest("/shorten", {
            long_url: urlInput.value,
        });
        urlObject = {
            long_url: response.long_url,
            short_url: response.short_url,
        };
        isShortened = checkResponse(response, urlFieldWrapper);
    } else {
        console.error(`Submit Btn Classes: ${submitBtnClasses}`);
    }

    if (isCodeSent) {
        emailInput.setAttribute("readonly", "true");
        codeFieldWrapper.style.display = "block";
        submitBtn.classList.replace("send-code-btn", "sign-in-btn");
        submitBtn.setAttribute("value", "Sign In");
    }

    if (isAuthenticated) {
        setTimeout(() => {
            window.location.reload();
        }, 4000);
    }

    if (isShortened) {
        setTimeout(() => {
            urlInput.value = "";
            createCardElement(urlObject["long_url"], urlObject["short_url"]);
        }, 1000);
    }
    submitBtn.removeAttribute("disabled");
});

// Set Current Year in Copyright Footer
const setFooter = () => {
    const element = document.getElementById("currentYear");
    const date = new Date();
    element.innerText = date.getFullYear();
};

window.addEventListener("DOMContentLoaded", () => {
    setFooter();
});
